lib/spack/spack/util/compression.py

```diff
@@ -233,22 +242,24 @@ def unarchive(archive_file):
         # record name of new archive so we can extract
         # and later clean up
         decomped_tarball = decompressor(archive_file)
-        if check_extension(decomped_tarball, "tar"):
-            # run tar on newly decomped archive
-            outfile = _system_untar(decomped_tarball)
-            # clean intermediate archive to mimic end result
-            # produced by one shot decomp/extraction
-            os.remove(decomped_tarball)
-            return outfile
-        return decomped_tarball
+        # run tar on newly decomped archive
+        outfile = _system_untar(decomped_tarball)
+        # clean intermediate archive to mimic end result
+        # produced by one shot decomp/extraction
+        os.remove(decomped_tarball)
+        return outfile
 
     return unarchive
 
 
 def _py_lzma(archive_file):
     """Returns path to decompressed .xz files
     Decompress lzma compressed .xz files via python lzma module"""
-    decompressed_file = os.path.basename(strip_extension(archive_file, "xz"))
+    if check_extension(archive_file, "txz"):
+        decompressed_file = os.path.basename(strip_extension(archive_file, "txz"))
+        decompressed_file += ".tar"
```

Comment:
Are there not similar issues for `.tgz`?

I assume an analogous change isn't necessary because of the changes in `_win_compressed_tarball_handler` (which no longer bails if the extension is not `.tar`); that also means that adding `decompressed_file += ".tar"` is unnecessary, although to be clear, I think stripping the `txz` extension is good.

I think other functions should copy the behavior of properly stripping the combined extensions (e.g. "tgz", "tbz"), since otherwise the intermediate decompressed (but not un-tar-ed) file retains the extension type, which could cause confusion when folks debug.

Reply:
> Are there not similar issues for .tgz?
On windows, this extension goes through _system_untar so it behaves different there.

> I assume an analogous change isn't necessary because of the changes in _win_compressed_tarball_handler (which no longer bails if the extension is not .tar); that also means that adding decompressed_file += ".tar" is unnecessary, although to be clear, I think stripping the txz extension is good.

Indeed, tests pass without .tar extension.